File: CS/CommonModule/CommonModule.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CommonModule.BusinessObjects;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.ExpressApp.DC.Xpo;
using DevExpress.ExpressApp.Security;
using DevExpress.ExpressApp.Security.Adapters;
using DevExpress.ExpressApp.Security.ClientServer;
using DevExpress.ExpressApp.Updating;
using DevExpress.ExpressApp.Xpo;

namespace CommonModule {
    public class Module : ModuleBase {
        public Module() {
            this.AdditionalExportedTypes.Add(typeof(DevExpress.Persistent.BaseImpl.ModelDifference));
            this.AdditionalExportedTypes.Add(typeof(DevExpress.Persistent.BaseImpl.ModelDifferenceAspect));
            this.AdditionalExportedTypes.Add(typeof(ApplicationUser));
            this.RequiredModuleTypes.Add(typeof(DevExpress.ExpressApp.SystemModule.SystemModule));
            this.RequiredModuleTypes.Add(typeof(DevExpress.ExpressApp.Security.SecurityModule));
        }
        private static readonly object lockObj = new object();
        // Here we will have a single instance, which is initialized only once during the application life cycle.
        private static XpoTypeInfoSource typeInfoSource = null;
        public override void Setup(XafApplication application) {
            base.Setup(application);
            application.CreateCustomObjectSpaceProvider += application_CreateCustomObjectSpaceProvider;
            application.ObjectSpaceCreated += Application_ObjectSpaceCreated;
        }
        private void Application_ObjectSpaceCreated(object sender, ObjectSpaceCreatedEventArgs e) {
            (e.ObjectSpace as CompositeObjectSpace)?.PopulateAdditionalObjectSpaces((XafApplication)sender);
        }
        void application_CreateCustomObjectSpaceProvider(object sender, CreateCustomObjectSpaceProviderEventArgs e) {
            XafApplication application = (XafApplication)sender;
            if(typeInfoSource == null) {
                lock(lockObj) {
                    if(typeInfoSource == null) {
                        typeInfoSource = new XpoTypeInfoSource((TypesInfo)application.TypesInfo,
                            typeof(DevExpress.Persistent.BaseImpl.ModelDifference),
                            typeof(DevExpress.Persistent.BaseImpl.ModelDifferenceAspect),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyUser),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyRole),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyTypePermissionObject),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyObjectPermissionsObject),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyMemberPermissionsObject),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyNavigationPermissionObject),
                            typeof(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyActionPermissionObject),
                            typeof(ApplicationUser)
                        );
                    }
                }
            }
            IObjectSpaceProvider objectSpaceProvider = new SecuredObjectSpaceProvider(
                (SecurityStrategyComplex)application.Security,
                    new ConnectionStringDataStoreProvider(ConfigurationManager.ConnectionStrings["ConnectionStringDatabaseCommon"].ConnectionString),
                    application.TypesInfo,
                    typeInfoSource, true
                );
            objectSpaceProvider.CheckCompatibilityType = CheckCompatibilityType.DatabaseSchema;
            e.ObjectSpaceProviders.Add(objectSpaceProvider);
            e.ObjectSpaceProviders.Add(new NonPersistentObjectSpaceProvider(application.TypesInfo, null));
            ((SecurityStrategyComplex)application.Security).RegisterXPOAdapterProviders(new SecurityPermissionsProviderDefault(application));
        }
        public override IEnumerable<ModuleUpdater> GetModuleUpdaters(IObjectSpace objectSpace, Version versionFromDB) {
            ModuleUpdater updater = new CommonModule.DatabaseUpdate.Updater(objectSpace, versionFromDB);
            return new ModuleUpdater[] { updater };
        }
    }
}

File: VB/CommonModule/CommonModule.vb
Imports System
Imports System.Collections.Generic
Imports System.Configuration
Imports CommonModule.BusinessObjects
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.ExpressApp.DC.Xpo
Imports DevExpress.ExpressApp.Security
Imports DevExpress.ExpressApp.Security.Adapters
Imports DevExpress.ExpressApp.Security.ClientServer
Imports DevExpress.ExpressApp.Updating
Imports DevExpress.ExpressApp.Xpo

Namespace CommonModule

    Public Class [Module]
        Inherits ModuleBase

        Public Sub New()
            Me.AdditionalExportedTypes.Add(GetType(DevExpress.Persistent.BaseImpl.ModelDifference))
            Me.AdditionalExportedTypes.Add(GetType(DevExpress.Persistent.BaseImpl.ModelDifferenceAspect))
            Me.AdditionalExportedTypes.Add(GetType(ApplicationUser))
            Me.RequiredModuleTypes.Add(GetType(DevExpress.ExpressApp.SystemModule.SystemModule))
            Me.RequiredModuleTypes.Add(GetType(DevExpress.ExpressApp.Security.SecurityModule))
        End Sub

        Private Shared ReadOnly lockObj As Object = New Object()

        ' Here we will have a single instance, which is initialized only once during the application life cycle.
        Private Shared typeInfoSource As XpoTypeInfoSource = Nothing

        Public Overrides Sub Setup(ByVal application As XafApplication)
            MyBase.Setup(application)
            application.CreateCustomObjectSpaceProvider += AddressOf application_CreateCustomObjectSpaceProvider
            TryCast(application.Security, SecurityStrategy)?.RegisterXPOAdapterProviders(New SecurityPermissionsProviderDefault(application))
            application.ObjectSpaceCreated += AddressOf Application_ObjectSpaceCreated
        End Sub

        Private Sub Application_ObjectSpaceCreated(ByVal sender As Object, ByVal e As ObjectSpaceCreatedEventArgs)
            TryCast(e.ObjectSpace, CompositeObjectSpace)?.PopulateAdditionalObjectSpaces(CType(sender, XafApplication))
        End Sub

        Private Sub application_CreateCustomObjectSpaceProvider(ByVal sender As Object, ByVal e As CreateCustomObjectSpaceProviderEventArgs)
            Dim application As XafApplication = CType(sender, XafApplication)
            If typeInfoSource Is Nothing Then
                SyncLock lockObj
                    If typeInfoSource Is Nothing Then
                        typeInfoSource = New XpoTypeInfoSource(CType(application.TypesInfo, TypesInfo), GetType(DevExpress.Persistent.BaseImpl.ModelDifference), GetType(DevExpress.Persistent.BaseImpl.ModelDifferenceAspect), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyUser), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyRole), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyTypePermissionObject), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyObjectPermissionsObject), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyMemberPermissionsObject), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyNavigationPermissionObject), GetType(DevExpress.Persistent.BaseImpl.PermissionPolicy.PermissionPolicyActionPermissionObject), GetType(ApplicationUser))
                    End If

                End SyncLock
            End If

            'IObjectSpaceProvider objectSpaceProvider = new XPObjectSpaceProvider(
            Dim objectSpaceProvider As IObjectSpaceProvider = New SecuredObjectSpaceProvider(CType(application.Security, SecurityStrategyComplex), New ConnectionStringDataStoreProvider(ConfigurationManager.ConnectionStrings("ConnectionStringDatabaseCommon").ConnectionString), application.TypesInfo, typeInfoSource, True)
            objectSpaceProvider.CheckCompatibilityType = CheckCompatibilityType.DatabaseSchema
            e.ObjectSpaceProviders.Add(objectSpaceProvider)
            e.ObjectSpaceProviders.Add(New NonPersistentObjectSpaceProvider(application.TypesInfo, Nothing))
        End Sub

        Public Overrides Function GetModuleUpdaters(ByVal objectSpace As IObjectSpace, ByVal versionFromDB As Version) As IEnumerable(Of ModuleUpdater)
            Dim updater As ModuleUpdater = New DatabaseUpdate.Updater(objectSpace, versionFromDB)
            Return New ModuleUpdater() {updater}
        End Function
    End Class
End Namespace
